### DISCUSSION TITLE
No target display found for L2 trim - please help

### DISCUSSION THREAD

**Discussion by Shao85** [quietvoid/dovi_tool#224] [Q&A] [answered]
Hello,
could you anybody advice with the conversion problem? Unfortunatelly, I'm not a programmer so, my debugging skills are pretty limited (have no clue how to use RUST_BACKTRACE etc...)

I have done these steps:

1. Generating 8.1 RPU binary from profile 7 using DDVT tool _(the mkv file has this HDR format: Dolby Vision, Version 1.0, dvhe.07.06, BL+EL+RPU, Blu-ray compatible / SMPTE ST 2086, HDR10 compatible)_

2. Using dovi_meta-0.1.0-x86_64-pc-windows-msvc for conversion of the binary file into a Dolby Vision XML metadata file for making some planned changes _(Metadata Version 2.0.5 XML for CMv2.9)_

3. Using dovi_tool-2.0.3-x86_64-pc-windows-msvc for conversion of the DV XML metadata file (so far without any changes) back into binary RPU _(8.1 profile as before)_

The last step simply doesn't works and I'm not sure where can be the root cause. Attached is 8.1 RPU; generated XML and snapshot from 3rd step.
[RPU_8.1.zip](https://github.com/quietvoid/dovi_tool/files/11034392/RPU_8.1.zip)
[metadata_8.1.zip](https://github.com/quietvoid/dovi_tool/files/11034409/metadata_8.1.zip)
![No target display found for L2 trim](https://user-images.githubusercontent.com/96745047/226757557-9008d6dc-8c0c-4cdf-bb66-87eb3b0a85d8.PNG)

Thanks for any help.

**Comment by quietvoid** [comment]
You should open an issue at [the dovi_meta repo](https://github.com/saindriches/dovi_meta).
It doesn't seem to be parsing all frames for L0 metadata, so the target displays are not all populated for L2 trims.

`dovi_tool` requires that all targets used in the trims are present in L0.

**Reply by Shao85** [reply]
Thanks. The case opened. Maybe will @saindriches find some time in a near future to fix the problem.

I have an idea for a new editing function. Maybe you both could tell me if is stupid or not. Let say I want to increase framerate of some given video which contains dynamic DV data. For the video interpolation I can use sw like Flowframes. Before changing the DV dynamic data is needed conversion from RPU to XML. Then must be edited(multiplied) _**In**_ and _**Duration**_  values for all the shots and the whole XML converted back to RPU binary.

![Shots info ](https://user-images.githubusercontent.com/96745047/227020264-f41d7765-c229-489d-8a53-78182a1a90c2.PNG)

**Comment by saindriches** [comment, marked as answer]
dovi_meta is for my personal use before, without considering more cases, and has some known issues. I'm not sure if I have time to fix them in recent days, for a workaround, you can manually add the missing `TargetDisplay`:
```
                <TargetDisplay level="0">
                  <ID>27</ID>
                  <Name>600-nits, P3, D65, ST.2084, Full</Name>
                  <Primaries>
                    <Red>0.68,0.32</Red>
                    <Green>0.265,0.69</Green>
                    <Blue>0.15,0.06</Blue>
                  </Primaries>
                  <WhitePoint>0.3127,0.329</WhitePoint>
                  <PeakBrightness>600</PeakBrightness>
                  <MinimumBrightness>0</MinimumBrightness>
                  <DiagonalSize>42</DiagonalSize>
                  <Encoding>pq</Encoding>
                  <BitDepth>16</BitDepth>
                  <ColorSpace>rgb</ColorSpace>
                  <SignalRange>computer</SignalRange>
                </TargetDisplay>
                <TargetDisplay level="0">
                  <ID>48</ID>
                  <Name>1000-nits, P3, D65, ST.2084, Full</Name>
                  <Primaries>
                    <Red>0.68,0.32</Red>
                    <Green>0.265,0.69</Green>
                    <Blue>0.15,0.06</Blue>
                  </Primaries>
                  <WhitePoint>0.3127,0.329</WhitePoint>
                  <PeakBrightness>1000</PeakBrightness>
                  <MinimumBrightness>0</MinimumBrightness>
                  <DiagonalSize>42</DiagonalSize>
                  <Encoding>pq</Encoding>
                  <BitDepth>16</BitDepth>
                  <ColorSpace>rgb</ColorSpace>
                  <SignalRange>computer</SignalRange>
                </TargetDisplay>
```

**Reply by Shao85** [reply]
Thanks! Adding of the missing **Target display** information helped. I can generate the RPU now. I just wonder what else could be missing as the generated RPU file is now about 1.6MB smaller compared to the original 8.1 file.

[metadata_8.1 (converted back from XML).zip](https://github.com/quietvoid/dovi_tool/files/11043848/metadata_8.1.converted.back.from.XML.zip)

**Reply by saindriches** [reply]
The Level 4 metadata in original RPU are not included in XML, they are generated during encoding. dovi_meta also discards Level 6 values by default, since they are not needed in Dolby Vision. You can keep Level 6 with `-6` flag in `convert` command.
